ksql-engine/src/main/java/io/confluent/ksql/function/UdfLoader.java

```diff
@@ -79,7 +79,7 @@
 
 
   @SuppressWarnings("OptionalUsedAsFieldOrParameterType")
-  UdfLoader(
+  public UdfLoader(
```

Comment:
Rather than make this internal constructor public, it would be better to add a new public constructor with just the bits needed.  This internal one requires stuff that no client of the class should need to pass - this constructor is for testing only.

I suggest:

```java
public UdfLoader(
      final MutableFunctionRegistry functionRegistry,
      final Optional<File> customUdfDirectory,
      final Optional<Metrics> metrics
  ) {
    this(
        functionRegistry,
        customUdfDirectory.orElse(new File("")),
        Thread.currentThread().getContextClassLoader(),
        customUdfDirectory
            .map(pluginDir -> (Predicate<String>) new Blacklist(
                new File(pluginDir, "resource-blacklist.txt")))
            .orElse(s -> false),
        new UdfCompiler(metrics),
        metrics,
        customUdfDirectory.isPresent()
    );
  }

  @SuppressWarnings("OptionalUsedAsFieldOrParameterType")
  @VisibleForTesting
  UdfLoader(
      final MutableFunctionRegistry functionRegistry,
      final File pluginDir,
      final ClassLoader parentClassLoader,
      final Predicate<String> blacklist,
      final UdfCompiler compiler,
      final Optional<Metrics> metrics,
      final boolean loadCustomerUdfs
  ) {
  ...
 }
```

This uses the presence of the `customPluginDirectory` to determine if plugins should be loaded.

The internal use of the constructor can also be switched to use this new constructor, to remove code duplication:

```java
public static UdfLoader newInstance(final KsqlConfig config,
      final MutableFunctionRegistry metaStore,
      final String ksqlInstallDir
  ) {
    final boolean loadCustomerUdfs = config.getBoolean(KsqlConfig.KSQL_ENABLE_UDFS);
    final boolean collectMetrics = config.getBoolean(KsqlConfig.KSQL_COLLECT_UDF_METRICS);
    final String extDirName = config.getString(KsqlConfig.KSQL_EXT_DIR);
    final Optional<File> pluginDir = loadCustomerUdfs
        ? Optional.of(KsqlConfig.DEFAULT_EXT_DIR.equals(extDirName)
        ? new File(ksqlInstallDir, extDirName)
        : new File(extDirName))
        : Optional.empty();

    final Optional<Metrics> metrics = collectMetrics
        ? Optional.of(MetricCollectors.getMetrics())
        : Optional.empty();

    if (config.getBoolean(KsqlConfig.KSQL_UDF_SECURITY_MANAGER_ENABLED)) {
      System.setSecurityManager(ExtensionSecurityManager.INSTANCE);
    }

    return new UdfLoader(metaStore, pluginDir, metrics);
  }
```

For extra brownie points you can drop the `loadCustomerUdfs` field completely by switching the `pluginDir` field to `Optional<File>`...